include/libdnf-cli/session.hpp

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright Contributors to the libdnf project.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2.1 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+
+#ifndef LIBDNF_CLI_SESSION_HPP
+#define LIBDNF_CLI_SESSION_HPP
+
+
+#include "argument_parser.hpp"
+
+
+namespace libdnf::cli::session {
+
+
+class Command;
+
+
+class Session {
+public:
+    /// @return Root command that represents the main program.
+    ///         The returned pointer must **not** be freed manually.
+    /// @since 5.0
+    Command * get_root_command() { return root_command.get(); }
+
+    /// Set `command` as the root command that represents the main program.
+    /// Session becomes owner of the `command`.
+    /// @since 5.0
+    void set_root_command(std::unique_ptr<Command> command) { root_command = std::move(command); }
+
+    /// @return Selected (sub)command that a user specified on the command line.
+    ///         The returned pointer must **not** be freed manually.
+    /// @since 5.0
+    Command * get_selected_command() { return selected_command; }
+
+    /// Set `command` as the selected (sub)command that a user specified on the command line.
+    /// We're only pointing to a command that is owned by the Session already.
+    /// @since 5.0
+    void set_selected_command(Command * command) { selected_command = command; }
+
+    /// @return The underlying argument parser.
+    /// @since 5.0
+    libdnf::cli::ArgumentParser & get_argument_parser() { return argument_parser; }
+
+private:
+    std::unique_ptr<Command> root_command;
+    Command * selected_command;
+    libdnf::cli::ArgumentParser argument_parser;
+};
+
+
+class Command {
+public:
+    explicit Command(Command & parent, std::string name);
+    explicit Command(Session & session);
+    virtual ~Command() = default;
+
+    /// Run the command. Must be overriden in inherited classes.
```

Comment:
Really "must" ? A default empty implementation is provided.
Another solution is to use pure virtual method. It "must" then be implemented in a derived class:
`virtual void run() = 0;`